Source/Filter/Filter.cpp

```diff
@@ -0,0 +1,256 @@
+                              +tmp(i-ix,j+iy,k-iz,scomp+n)
+                              +tmp(i+ix,j+iy,k-iz,scomp+n)
+                              +tmp(i-ix,j-iy,k+iz,scomp+n)
+                              +tmp(i+ix,j-iy,k+iz,scomp+n)
+                              +tmp(i-ix,j+iy,k+iz,scomp+n)
+                              +tmp(i+ix,j+iy,k+iz,scomp+n));
+#else
+                    d += sss*( tmp(i-ix,j-iy,k,scomp+n)
+                              +tmp(i+ix,j-iy,k,scomp+n)
+                              +tmp(i-ix,j+iy,k,scomp+n)
+                              +tmp(i+ix,j+iy,k,scomp+n));
+#endif
+                }
+            }
+        }
+
+        dst(i,j,k,dcomp+n) = d;
+    });
+}
+
+#else
+
+/* \brief Apply stencil on MultiFab (CPU version, 2D/3D).
+ * \param dstmf Destination MultiFab
+ * \param srcmf source MultiFab
+ * \param scomp first component of srcmf on which the filter is applied
+ * \param dcomp first component of dstmf on which the filter is applied
+ * \param ncomp Number of components on which the filter is applied.
+ */
+void
+Filter::ApplyStencil (MultiFab& dstmf, const MultiFab& srcmf, int scomp, int dcomp, int ncomp)
+{
+    BL_PROFILE("BilinearFilter::ApplyStencil()");
+    ncomp = std::min(ncomp, srcmf.nComp());
+#ifdef _OPENMP
+#pragma omp parallel
+#endif
+    {
+        FArrayBox tmpfab;
+        for (MFIter mfi(dstmf,true); mfi.isValid(); ++mfi){
+            const auto& srcfab = srcmf[mfi];
+            auto& dstfab = dstmf[mfi];
+            const Box& tbx = mfi.growntilebox();
+            const Box& gbx = amrex::grow(tbx,stencil_length_each_dir-1);
+            // tmpfab has enough ghost cells for the stencil
+            tmpfab.resize(gbx,ncomp);
+            tmpfab.setVal(0.0, gbx, 0, ncomp);
+            // Copy values in srcfab into tmpfab
+            const Box& ibx = gbx & srcfab.box();
+            tmpfab.copy(srcfab, ibx, scomp, ibx, 0, ncomp);
+            // Apply filter
+            DoFilter(tbx, tmpfab.array(), dstfab.array(), 0, dcomp, ncomp);
+        }
+    }
+}
+
+/* \brief Apply stencil on FArrayBox (CPU version, 2D/3D).
+ * \param dstfab Destination FArrayBox
+ * \param srcmf source FArrayBox
+ * \param tbx Grown box on which srcfab is defined.
+ * \param scomp first component of srcfab on which the filter is applied
+ * \param dcomp first component of dstfab on which the filter is applied
+ * \param ncomp Number of components on which the filter is applied.
+ */
+void
+Filter::ApplyStencil (FArrayBox& dstfab, const FArrayBox& srcfab, 
+                      const Box& tbx, int scomp, int dcomp, int ncomp)
+{
+    BL_PROFILE("BilinearFilter::ApplyStencil(FArrayBox)");
+    ncomp = std::min(ncomp, srcfab.nComp());
+    FArrayBox tmpfab;
+    const Box& gbx = amrex::grow(tbx,stencil_length_each_dir-1);
+    // tmpfab has enough ghost cells for the stencil
+    tmpfab.resize(gbx,ncomp);
+    tmpfab.setVal(0.0, gbx, 0, ncomp);
+    // Copy values in srcfab into tmpfab
+    const Box& ibx = gbx;
```

Comment:
@WeiqunZhang in the `MultiFab` version of ApplyStencil (just above), this line performs a box intersection:
```
const Box& ibx = gbx & srcfab.box();
```
I think this is not required for this routine, that works on a single `FArrayBox`. Is that correct? I am not too sure.

Reply:
I think it is required.